packages/styled-components/src/test/types.tsx

```diff
@@ -149,15 +150,153 @@ const DivWithProps = styled.div<{ waz: number }>`
+  // Should allow hyphenated props
+  "data-test": 42,
+})``;
+
+const AttrRequiredTest2 = styled(DivWithRequiredProps).attrs({
+  // @ts-expect-error foo must be a number
+  foo: "not a number"
+})``;
+
+<AttrRequiredTest foo={42} bar="bar" />;
+// @ts-expect-error foo and bar are required props
+<AttrRequiredTest bar="bar" />;
+
+/** Intrinsic props and ref are being incorrectly types when using `as`
+ * https://github.com/styled-components/styled-components/issues/3800#issuecomment-1548941843
+ */
+const Text = styled.p``;
+// @ts-expect-error Can't treat paragraph element as label element
+<Text onCopy={(e: React.ClipboardEvent<HTMLLabelElement>) => {}} />;
+<Text
+  as="label"
+  ref={(el: HTMLLabelElement | null) => {}}
+  onCopy={(e: React.ClipboardEvent<HTMLLabelElement>) => {}}
+/>;
+<Text as="label" ref={(e: HTMLLabelElement | null) => {}} />;
+// @ts-expect-error Should now be a label element
+<Text as="label" ref={(e: HTMLParagraphElement | null) => {}} />;
+
+const AttrObjectAsLabel = styled(Text).attrs({ as: 'label' })``;
+<AttrObjectAsLabel
+  ref={(el: HTMLLabelElement | null) => {}}
+  onCopy={(e: React.ClipboardEvent<HTMLLabelElement>) => {}}
+/>;
+
+const AttrFunctionAsLabel = styled(Text).attrs(() => ({ as: 'label' }))``;
+<AttrFunctionAsLabel
+  ref={(el: HTMLLabelElement | null) => {}}
+  onCopy={(e: React.ClipboardEvent<HTMLLabelElement>) => {}}
+/>;
+
+/**
+ * Using IStyledComponent as the casted type of a functional component won't retain intrinsic prop types once styled
+ * https://github.com/styled-components/styled-components/issues/3800#issuecomment-1548941843
+ */
+const StyledDiv = styled.div``;
+
+const CustomComponent = (({ ...props }) => {
+  return <StyledDiv {...props} />;
+}) as IStyledComponent<'web', 'div'>;
+
+const StyledCustomComponent = styled(CustomComponent)``;
+
+<StyledCustomComponent className="class" />;
+
+// Performance test
+interface VeryLargeUnionProps {
+  foo: VeryLargeUnionType; // Comment out this line to compare performance
+}
+
+const UnstyledComponentVeryLargeUnion = (_props: VeryLargeUnionProps) => {
+  return <></>;
+};
+
+const StyledComponentVeryLargeUnion = styled(UnstyledComponentVeryLargeUnion)`
+  // Multiple template strings helps illustrate possible perf issue
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+  color: ${props => props.theme.waz};
+`;
+
+const AttrFunctionRequiredTest1 = styled(DivWithRequiredProps).attrs(props => ({
+  // Should not have to provide foo within attrs
+  bar: "hello",
+  // Should allow additional props
+  newProp: 42,
+  // Should allow hyphenated props
+  "data-foo": 42,
+  // Props should have inherited props
+  foo2: props.foo?.toPrecision(0),
```

Comment:
Allowing `foo2` is a bug, we don't want to allow unknown props to be returned from `attrs`.

Reply:
I noticed v5 type defs allowed it, so I thought maybe it was a feature.  I'll update it to restrict it so only unknown data attributes are allowed.

Reply:
@probablyup I've tried a bunch of ways to prevent unknown props within the attrs, but it doesn't play nicely with TypeScript's normal behavior of generic argument duck checking.  

I should note, this sort of thing isn't typically checked in other simple scenarios when it comes to return types, so perhaps we shouldn't expect TS to error on this situation.  It may lead to some unexpected side effects to have it be _too_ strict.

For example:
```ts
interface MyProps {
  foo: number;
}

type MyPropsFactory = () => MyProps;

// Error: Object literal may only specify known properties, and 'bar' does not exist in type 'MyProps'
const props: MyProps = { foo: 42, bar: 52 };

// No error
const propFactory: MyPropsFactory = () => ({ foo: 42, bar: 52 });
```

Reply:
> perhaps we shouldn't expect TS to error on this situation

The problem is lack of this is what leads to subtle errors due to typos. If you misspell something by a character it'll still take the prop and not alert you that it's buggy.

Reply:
Alrighty!  I managed to add these restrictions, but boy was it difficult.  Like, really difficult.

One tradeoff to note -- I think it slowed down the compilation a bit.  `tsc` of the styled-components folder went from 8sec to 12sec.  Not a huge amount, but just felt like I would note it.